src/ContextEngine/Contexts/PersistentContext.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+namespace OpenDialogAi\ContextEngine\Contexts;
+
+use Ds\Map;
+use Illuminate\Support\Facades\Log;
+use OpenDialogAi\AttributeEngine\AttributeBag\HasAttributesTrait;
+use OpenDialogAi\AttributeEngine\Contracts\Attribute;
+use OpenDialogAi\AttributeEngine\Exceptions\AttributeDoesNotExistException;
+use OpenDialogAi\ContextEngine\Contracts\ContextDataClient;
+use OpenDialogAi\ContextEngine\Exceptions\UnableToLoadAttributeFromPersistentStorageException;
+
+class PersistentContext extends AbstractContext
+{
+
+    protected $persistentAttributes = [];
+
+    protected ContextDataClient $dataClient;
+
+    public function __construct($id, ContextDataClient $dataClient)
+    {
+        parent::__construct($id);
+        $this->dataClient = $dataClient;
+    }
+
+    /**
+     * If we are dealing with an attribute that is not persistent try to retrieve as usual, alternatively
+     * if it is a persistent attributet has not already been retrieved or if we are meant to refresh its
+     * value we access through the data client.
+     *
+     * @param string $attributeName
+     * @param bool $refresh
+     * @return Attribute
+     */
+    public function getAttribute(string $attributeName, bool $refresh = false): Attribute
+    {
+        if (!$this->isPersistentAttribute($attributeName)) {
+            return parent::getAttribute($attributeName);
+        }
+
+        if ($this->hasAttribute($attributeName) && !$refresh) {
+            return parent::getAttribute($attributeName);
+        } elseif (($this->hasAttribute($attributeName) && $refresh) || (!$this->hasAttribute($attributeName))) {
+            try {
+                $this->dataClient->loadAttribute($attributeName);
```

Comment:
Is there a missing bit here where, after loading the attribute, it is added to the internal attribute map and returned? Something like:
```suggestion
                $attribute = $this->dataClient->loadAttribute($attributeName);
                this->attributes->put($attribute->getId(), $attribute);
                return $attribute;
```

Reply:
Good catch :-) added a fix for this.